server/src/main/java/io/crate/replication/logical/action/ReplayChangesAction.java

```diff
@@ -86,63 +96,103 @@ protected ReplicationResponse newResponseInstance(StreamInput in) throws IOExcep
         }
 
         @Override
-        protected void shardOperationOnPrimary(Request request,
-                                               IndexShard primary,
-                                               ActionListener<PrimaryResult<Request, ReplicationResponse>> listener) {
-            ActionListener.completeWith(
-                listener,
-                () -> new WritePrimaryResult<>(
+        protected void shardOperationOnPrimary(Request request, IndexShard primary, ActionListener<PrimaryResult<Request, ReplicationResponse>> listener) {
+            performOnPrimary(primary, request, new ArrayList<>(), 0, results -> {
```

Comment:
The downside of this approach is that all `Engine.Result` results are kept in memory until they are processed, so this operation will increase memory consumption temporarily.